### ISSUE TITLE
[SignalR] Server stops sending ping message to the client

### ISSUE THREAD

**Issue by colinblaise** [dotnet/aspnetcore#47327]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

The server stops sending ping messages to the client, which causes the connection to abort.

I tried to dig into how the ping message might be getting blocked by combing through https://github.com/dotnet/aspnetcore/blob/bec278eabea54f63da15e10e654bdfa4168a2479/src/SignalR/server/Core/src/HubConnectionContext.cs

```c#
 private ValueTask TryWritePingAsync()
    {
        // Don't wait for the lock, if it returns false that means someone wrote to the connection
        // and we don't need to send a ping anymore
        if (!_writeLock.Wait(0))
        {
            return default;
        }

        // TODO: cancel?
        return new ValueTask(TryWritePingSlowAsync());
    }

private async Task TryWritePingSlowAsync()
    {
        try
        {
            if (_connectionAborted)
            {
                return;
            }

            await _connectionContext.Transport.Output.WriteAsync(_cachedPingMessage);

            Log.SentPing(_logger);
        }
        catch (Exception ex)
        {
            CloseException = ex;
            Log.FailedWritingMessage(_logger, ex);
            AbortAllowReconnect();
        }
        finally
        {
            _writeLock.Release();
        }
    }
```

I have a feeling it is getting hung up on 
```c#
await _connectionContext.Transport.Output.WriteAsync(_cachedPingMessage);
```

But I don't know for sure.  

Are there any steps I can take to help narrow down the cause?

So far, I've failed to reproduce this bug in test project

### Expected Behavior

The server should never stop sending ping messages to the client.  In the event it cannot, it would be nice if there were logs indicating why the ping message isn't being sent.

### .NET Version

7.0.102

### Anything else?

client: @microsoft/signalr@7.0.2

This bug has been occurring in our project for over a year now.

**Comment by BrennanConroy** [comment]
> Are there any steps I can take to help narrow down the cause?

Turn on logging and gather network traces.
https://learn.microsoft.com/aspnet/core/signalr/diagnostics?view=aspnetcore-7.0

Also, a brief description of how your client and server interact would be helpful.

**Comment by colinblaise** [comment]
I have trace logging enabled on the client and server.

![image](https://user-images.githubusercontent.com/6540262/226472272-d65e5f6e-e421-449b-86e6-46059c5f3c89.png)

I'm able to see threads that are stuck waiting for the semaphore to release.  I'm pretty sure the server stops sending ping events `if (!_writeLock.Wait(0))`. is returning true when it's locked.  

The client jobs a connection group and starts receiving a lot of data.  Eventually signal stops sending data and stops sending ping events, which causes the connection to abort.

**Comment by BrennanConroy** [comment]
> I'm pretty sure the server stops sending ping events `if (!_writeLock.Wait(0))`. is returning true when it's locked.

Sure, if the server is sending data then pings aren't needed so it will skip sending it and check again after X seconds.

> I'm able to see threads that are stuck waiting for the semaphore to release.

Is the server sending faster than the client can process? What log do you get when the connection closes?

**Comment by colinblaise** [comment]
I agree with the design principle, however, I'm not receiving data and since `writeLock` is held it isn't sending the ping either.

Here is the error I receive when the keep alive interval lapses:

![image](https://user-images.githubusercontent.com/6540262/226477597-e7260dda-69d7-4648-b4f1-73ae73ebb5fe.png)

**Comment by colinblaise** [comment]
I have a bunch of server -> client methods that aren't actually being awaited.  I wonder if this is somehow causing a deadlock.

![image](https://user-images.githubusercontent.com/6540262/226478097-f2279e26-296e-4f9f-b20a-bbd09bcad916.png)

**Comment by colinblaise** [comment]
It's looking like this is the problem.

Having synchronous methods call a method that has an internal semaphore is a recipe for deadlock disaster.

After updated it to this, I haven't been able to reproduce the issue.

![image](https://user-images.githubusercontent.com/6540262/226482582-71d18b6e-b8d3-45b6-97f2-a78cf791c503.png)
